decred/decred/dcr/account.py

```diff
@@ -831,19 +1088,169 @@ def calcBalance(self, tipHeight=None):
+        for addr in addrs:
+            if addr in self.txs:
+                txids.extend(self.txs[addr])
+
+        for txid in set(txids):
+            # no need to check confirmed revocations and votes
+            if txid in votedTickets or txid in revokedTickets:
+                if self.ticketDB[txid].tinfo.status != "unconfirmed":
+                    continue
+            if txid in tickets:
+                # no need to check live tickets that we know about and have been mined
+                if txid in unspentTickets and unspentTickets[txid].isLiveTicket():
+                    continue
+                log.info("found unspent ticket: {}".format(txid))
+                ticket = tickets[txid]
+                # Estimate txs fee. All ticket purchases for pools are about 540
+                # bytes.
+                ticket.tinfo.purchaseTxFee = txscript.calcMinRequiredTxRelayFee(
+                    self.blockchain.relayFee(), 540
+                )
+                self.ticketDB[txid] = unspentTickets[txid] = ticket
+                continue
+
+            tx = self.blockchain.tx(txid)
+            # Vout 0 determines whether this is a ticket.
+            txout = tx.txOut[0]
+            if txscript.isStakeSubmissionScript(0, txout.pkScript):
+                tinfo = self.spentTicketInfo(txid)
+                # Continue on a still unconfirmed vote/transaction.
+                if not tinfo:
+                    continue
+                height = tinfo.purchaseBlock.height
+                ticket = UTXO.ticketFromTx(tx, self.net, height, tinfo)
+
+                # Set the transaction fee for the original purchase.
+                _, tinfo.purchaseTxFee = TicketInfo.fees(tx)
+
+                if tinfo.vote:
+                    log.info("found vote for ticket: {}".format(txid))
+                    vote = self.blockchain.tx(reversed(tinfo.vote).hex())
+                    tinfo.poolFee, tinfo.spendTxFee = TicketInfo.fees(vote)
+                    tinfo.stakebase = vote.txIn[0].valueIn
+                    self.ticketDB[txid] = votedTickets[txid] = ticket
+                    continue
+
+                elif tinfo.revocation:
+                    log.info("found revocation for ticket: {}".format(txid))
+                    revocation = self.blockchain.tx(reversed(tinfo.revocation).hex())
+                    tinfo.poolFee, tinfo.spendTxFee = TicketInfo.fees(revocation)
+                    self.ticketDB[txid] = revokedTickets[txid] = ticket
+
+        # Because unspentTickets deal with immature tickets and those in the
+        # mempool, it's possible this ticket never got mined. Remove such
+        # tickets from the database.
+        for txid in self.unspentTickets():
+            if txid not in tickets:
+                # If this is an unconfirmed spent ticket, leave it. It will be
+                # updated when confirmed.
+                if self.blockchain.existsTx(txid):
+                    continue
```

Comment:
`existsTx` always required an API call. Is this needed for all unspent tickets, or just mempool tickets?

Reply:
I'm not entirely sure how reorgs go down, but you could have a ticket mined in a block and that block disappear in a reorg correct?